packages/alfa-rules/src/common/diagnostic/contrast.ts

```diff
@@ -125,9 +135,58 @@ export namespace Contrast {
   export namespace Pairing {
     export interface JSON {
       [key: string]: json.JSON;
-      foreground: RGB.JSON;
-      background: RGB.JSON;
+      color1: Color.JSON;
+      color2: Color.JSON;
       contrast: number;
     }
   }
+
+  type Name = "container" | "link" | "foreground" | "background";
+  export class Color implements Equatable, Serializable, Hashable {
```

Comment:
This could have a `<N extends Name = Name>` type parameter, so that the type system knows which kind of colour we use. We could even go as far as adding a `<N extends ["container", "link"] | ["foreground", "background"]>` parameter to `Pairing` (and use types `N[0]`, `N[1]`) to ensure we only build meaningful pairings (and we unpack them correctly).

There could be 
```
Pairing.from<N extends ["container", "link"] | ["foreground", "background"]>(color1: [N[0], number], color2: [N[1], number], contrast: number) {
  return Pairing.of(Color.of(...color1), Color.of(...color2), contrast)
}
```

`ElementDistinguishable` could then make sure that it's pairing list is correctly typed (`Pairing<"container", "link">`).